src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/IncludeLocalesPlugin.java

```diff
@@ -270,26 +269,47 @@ private List<String> includeLocaleFiles(String localeStr) {
             .toList();
     }
 
-    private boolean stripUnsupportedLocales(byte[] bytes, ClassReader cr) {
-        boolean[] modified = new boolean[1];
-
-        IntStream.range(1, cr.getItemCount())
-            .map(item -> cr.getItem(item))
-            .forEach(itemIndex -> {
-                if (bytes[itemIndex - 1] == 1 &&         // UTF-8
-                    bytes[itemIndex + 2] == (byte)' ') { // fast check for leading space
-                    int length = cr.readUnsignedShort(itemIndex);
-                    byte[] b = new byte[length];
-                    System.arraycopy(bytes, itemIndex + 2, b, 0, length);
-                    if (filterOutUnsupportedTags(b)) {
-                        // copy back
-                        System.arraycopy(b, 0, bytes, itemIndex + 2, length);
-                        modified[0] = true;
+    private boolean stripUnsupportedLocales(byte[] bytes) {
+        boolean modified = false;
+        // scan CP entries directly to read the bytes of UTF8 entries and
+        // patch in place with unsupported locale tags stripped
+        IntUnaryOperator readU2 = p -> ((bytes[p] & 0xff) << 8) + (bytes[p + 1] & 0xff);
+        int cpLength = readU2.applyAsInt(8);
+        int offset = 10;
+        for (int cpSlot=1; cpSlot<cpLength; cpSlot++) {
+            switch (bytes[offset]) { //entry tag
+                case TAG_UTF8 -> {
+                    int length = readU2.applyAsInt(offset + 1);
+                    if (bytes[offset + 3] == (byte)' ') { // fast check for leading space
+                        byte[] b = new byte[length];
+                        System.arraycopy(bytes, offset + 3, b, 0, length);
+                        if (filterOutUnsupportedTags(b)) {
+                            // copy back
+                            System.arraycopy(b, 0, bytes, offset + 3, length);
+                            modified = true;
+                        }
                     }
+                    offset += 3 + length;
                 }
-            });
-
-        return modified[0];
+                case TAG_CLASS,
+                     TAG_STRING,
+                     TAG_METHODTYPE,
+                     TAG_MODULE,
+                     TAG_PACKAGE -> offset += 3;
+                case TAG_METHODHANDLE -> offset += 4;
+                case TAG_INTEGER,
+                     TAG_FLOAT,
+                     TAG_FIELDREF,
+                     TAG_METHODREF,
+                     TAG_INTERFACEMETHODREF,
+                     TAG_NAMEANDTYPE,
+                     TAG_CONSTANTDYNAMIC,
+                     TAG_INVOKEDYNAMIC -> offset += 5;
+                case TAG_LONG,
+                     TAG_DOUBLE -> {offset += 9; cpSlot++;} //additional slot for double and long entries
```

Comment:
should this add the default case and throw to prepare for future change?

Reply:
yes, fixed, thanks.

Reply:
@ExE-Boss I've applied your proposed exception message construction, thanks.